lib/virtio_mmio/virtio_mmio.c

```diff
@@ -391,10 +480,25 @@ struct virtqueue *virtio_mmio_setup_virtqueue(struct virtio_device *vdev,
 		metal_assert(maxq != 0);
 		metal_assert(maxq >= vq->vq_nentries);
 		virtio_mmio_write32(vdev, VIRTIO_MMIO_QUEUE_NUM, vq->vq_nentries);
+#if defined(HVL_VIRTIO)
+		if (vmdev && (vmdev->hvl_mode == 1)) {
+			virtio_mmio_hvl_ipi();
+		}
+#endif
 		virtio_mmio_write32(vdev, VIRTIO_MMIO_QUEUE_ALIGN, 4096);
 		virtio_mmio_write32(vdev, VIRTIO_MMIO_QUEUE_PFN,
 				    ((uintptr_t)metal_io_virt_to_phys(vq->shm_io,
 				    (char *)vq->vq_ring.desc)) / 4096);
+#if defined(HVL_VIRTIO)
+		if (vmdev && (vmdev->hvl_mode == 1)) {
+			virtio_mmio_hvl_ipi();
+			if (virtio_mmio_hvl_wait_cfg(vdev, VIRTIO_MMIO_QUEUE_PFN,
```

Comment:
For devices which use multiple virtqueues, the same registers are used to provide the information for all the queues. The workflow is to select the queue (VIRTIO_MMIO_QUEUE_SEL), followed by VIRTIO_MMIO_QUEUE_PFN to set the queue address and then wait for ACK from the PMM. This repeats for the number of virtqueues. DRIVER_OK follows at the end of this pseudo-negotiation sequence. Introducing a wait may work, although it will complicate ISR processing (to be considered for potential enhancement).

Reply:
Yes, VIRTIO_MMIO_HVL_CFG_ACK is used by the physical machine monitor (PMM) to signal it has processed the configuration request. It is specific to hypervisorless virtio.

Reply:
Please you add a comment that specify that here it is a derogation of the virtio protocol for HVL